### ISSUE TITLE
reportUndefinedVariable when method references itself

### ISSUE THREAD

**Issue by AllanDaemon** [microsoft/pyright#1620]
The following code works fine in pylance (no errors):

```python
def func(self):
	print(func)
```
But the following 3 codes give an error `"func" is not defined Pylance(reportUndefinedVariable)`, but it shouldn't:

```python
class A:
	def func(self):
		print(func) # "func" is not defined Pylance(reportUndefinedVariable) [3,9]
```

```python
class A:
	@staticmethod
	def func():
		print(func) # "func" is not defined Pylance(reportUndefinedVariable) [4,9]
```

```python
class A:
	@classmethod
	def func(cls):
		print(func) # "func" is not defined Pylance(reportUndefinedVariable) [4,9]
```

You could use `self.func` or `cls.func`, but this is still valid and should not raise an error, just like the first example (the function instead of method).

Here is the error copied as JSON:
```json
{
	"resource": "/home/user/bug.py",
	"owner": "_generated_diagnostic_collection_name_#3",
	"code": {
		"value": "reportUndefinedVariable",
		"target": {
			"$mid": 1,
			"external": "https://github.com/microsoft/pyright/blob/master/docs/configuration.md",
			"path": "/microsoft/pyright/blob/master/docs/configuration.md",
			"scheme": "https",
			"authority": "github.com"
		}
	},
	"severity": 8,
	"message": "\"func\" is not defined",
	"source": "Pylance",
	"startLineNumber": 3,
	"startColumn": 9,
	"endLineNumber": 3,
	"endColumn": 13
}
```

### VS Code Version
```yaml
Version: 1.54.1
Commit: f30a9b73e8ffc278e71575118b6bf568f04587c8
Date: 2021-03-04T22:38:50.094Z
Electron: 11.3.0
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Linux x64 5.10.19-200.fc33.x86_64
```
### Extension version
`Pylance ms-python.vscode-pylance preview v2021.3.1`

**Comment by erictraut** [comment]
Pyright is doing the right thing here.

All three of these generate runtime exceptions if you attempt to call the function. Try running the following in the Python interpreter.

```python
class A:
	def func(self):
		print(func)

class B:
	@staticmethod
	def func():
		print(func)

class C:
	@classmethod
	def func(cls):
		print(func)

A().func() # Generates exception
B.func() # Generates exception
C.func() # Generates exception
```

**Comment by AllanDaemon** [comment]
I see. I missed that part of the testing, sorry.
